### PR TITLE
Add builtin integer type to the surface language

### PR DESCRIPTION
**Pull request by paulcadman**
This PR adds a builtin integer type to the surface language that is compiled to the backend integer type.

## Inductive definition

The `Int` type is defined in the standard library as:

```
builtin int
type Int :=
  | --- ofNat n represents the integer n
    ofNat : Nat -> Int
  | --- negSuc n represents the integer -(n + 1)
    negSuc : Nat -> Int;
```

## New builtin functions defined in the standard library

```
intToString : Int -> String;
+ : Int -> Int -> Int;
neg : Int -> Int;
* : Int -> Int -> Int;
- : Int -> Int -> Int;
div : Int -> Int -> Int;
mod : Int -> Int -> Int;

== : Int -> Int -> Bool;
<= : Int -> Int -> Bool;
< : Int -> Int -> Bool;
```

Additional builtins required in the definition of the other builtins:

```
negNat : Nat -> Int;
intSubNat : Nat -> Nat -> Int;
nonNeg : Int -> Bool;
```

## REPL types of literals

In the REPL, non-negative integer literals have the inferred type `Nat`, negative integer literals have the inferred type `Int`. 

```
Stdlib.Prelude> :t 1
Nat
Stdlib.Prelude> :t -1
Int
:t let x : Int := 1 in x
Int
```

## The standard library Prelude

The definitions of `*`, `+`, `div` and `mod` are not exported from the standard library prelude as these would conflict with the definitions from `Stdlib.Data.Nat`.

Stdlib.Prelude
```
open import Stdlib.Data.Int hiding {+;*;div;mod} public;
```

* Closes https://github.com/anoma/juvix/issues/1679
* Closes https://github.com/anoma/juvix/issues/1984

### CODE REVIEW

**Comment by lukaszcz** [comment]
> > In the REPL:
> > ```
> > Stdlib.Prelude> :type div 4 -2
> > Nat
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The operations on Nats can somehow be used with integers. It shouldn't be possible or the result should have type Int.
> 
> Thanks for catching this. I've now fixed it and added a negative type checking test.
> 
> The REPL now reports:
> 
> ```
> Stdlib.Prelude> div 1 -2
> /<repl>:1:8-9: error:
> The expression -2 has type:
>   Int
> but is expected to have type:
>   Nat
> ```

This doesn't solve the problem completely:
```
Stdlib.Prelude> let f : String -> Nat := \{_ := 0} in f 2
0
```
Should give an error that `2` is expected to have type String. I guess one also needs to check if the inferred type is actually a subtype of the hint type in src/Juvix/Compiler/Internal/Translation/FromInternal/Analysis/TypeChecking/Checker.hs:626.

File: src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            Just BuiltinIntDiv -> f (\info x y -> mkBuiltinApp info OpIntDiv [x, y])
+            Just BuiltinIntMod -> f (\info x y -> mkBuiltinApp info OpIntMod [x, y])
+            Just BuiltinIntSub -> f (\info x y -> mkBuiltinApp info OpIntSub [x, y])
+            Just BuiltinIntLe -> f (\info x y -> mkBuiltinApp info OpIntLe [x, y])
+            Just BuiltinIntLt -> f (\info x y -> mkBuiltinApp info OpIntLt [x, y])
+            _ -> node
+
+    convertSingleArgIdentApp :: Node -> Node -> Info -> Symbol -> Node
+    convertSingleArgIdentApp node l info sym =
+      let ii = fromJust $ HashMap.lookup sym (tab ^. infoIdentifiers)
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs:130]: Use `lookupIdentifierInfo`

File: src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+                ( \g ->
+                    mkLet' mkTypeInteger' l $
+                      mkLambda' mkTypeInteger' $
+                        g info (mkVar' 1) (mkVar' 0)
+                )
+                sym
+
+    convertSingleArgIdent :: Node -> Info -> Symbol -> Node
+    convertSingleArgIdent node info sym =
+      let ii = fromJust $ HashMap.lookup sym (tab ^. infoIdentifiers)
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs:148]: Use `lookupIdentifierInfo`

File: src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        recur [] $ convertIdentApp node (\g -> g _identInfo l r) _identSymbol
+      NApp (App _ (NIdt (Ident {..})) l) ->
+        recur [] $ convertSingleArgIdentApp node l _identInfo _identSymbol
+      NIdt (Ident {..})
+        | Just _identSymbol == tab ^. infoLiteralIntToInt ->
+            mkLambda' mkTypeInteger' (mkVar' 0)
+      NIdt (Ident {..}) ->
+        recur [] $ convertSingleArgIdent node _identInfo _identSymbol
+      NCtr (Constr {..}) ->
+        let ci = fromJust $ HashMap.lookup _constrTag (tab ^. infoConstructors)
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs:29]: Use `lookupConstructorInfo`

**Comment by paulcadman** [comment]
> This doesn't solve the problem completely:
> 
> ```
> Stdlib.Prelude> let f : String -> Nat := \{_ := 0} in f 2
> 0
> ```
> 
> Should give an error that `2` is expected to have type String. I guess one also needs to check if the inferred type is actually a subtype of the hint type in src/Juvix/Compiler/Internal/Translation/FromInternal/Analysis/TypeChecking/Checker.hs:626.

Thank you - I've tried to solve the problem again.

File: src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    negSucConv n =
+      mkBuiltinApp'
+        OpIntSub
+        [ mkConstant' (ConstInteger 0),
+          mkBuiltinApp' OpIntAdd [n, mkConstant' (ConstInteger 1)]
+        ]
+
+    convertIdentApp :: Node -> ((Info -> Node -> Node -> Node) -> Node) -> Symbol -> Node
+    convertIdentApp node f sym =
+      let ii = fromJust $ HashMap.lookup sym (tab ^. infoIdentifiers)
```
[the remark this answers is not in the document]
Comment (paulcadman) [reply src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs]: done in bbbcec05dd1588510

File: src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs (hunk 1 above)
Comment (paulcadman) [reply src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs]: done in bbbcec05dd1588510

File: src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+             in mkIf
+                  _caseInfo
+                  boolSym
+                  (mkBuiltinApp' OpIntLe [mkConstant' (ConstInteger 0), cv])
+                  (mkLet mempty (binder' ofNatBranch) cv (mkBody ofNatBranch))
+                  (mkLet mempty (binder' negSucBranch) (negSucConv cv) (mkBody negSucBranch))
+
+          builtinCtor :: CaseBranch -> BuiltinIntCtor
+          builtinCtor CaseBranch {..} =
+            let ci = fromJust $ HashMap.lookup _caseBranchTag (tab ^. infoConstructors)
```
[the remark this answers is not in the document]
Comment (paulcadman) [reply src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs]: done in bbbcec05dd1588510

File: src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs (hunk 3 above)
Comment (paulcadman) [reply src/Juvix/Compiler/Core/Transformation/IntToPrimInt.hs]: done in bbbcec05dd1588510

**Review by lukaszcz — APPROVED** [review]

**Pull request merged**